### PR TITLE
libressl_fix

### LINKED ISSUE #8796: Build error building against libreseel-dev
##### Bug report summary

Getting the following error while trying to install netdata in Alpine, I followed the steps in the official [guide](https://learn.netdata.cloud/docs/agent/packaging/installer/methods/alpine).

```
libnetdata/socket/security.c: In function 'tls_select_version':
libnetdata/socket/security.c:71:16: error: 'TLS1_3_VERSION' undeclared (first use in this function); did you mean 'TLS1_2_VERSION'?
   71 |         return TLS1_3_VERSION;
      |                ^~~~~~~~~~~~~~
      |                TLS1_2_VERSION
libnetdata/socket/security.c:71:16: note: each undeclared identifier is reported only once for each function it appears in
libnetdata/socket/security.c:74:12: error: 'TLS_MAX_VERSION' undeclared (first use in this function); did you mean 'TLS1_2_VERSION'?
   74 |     return TLS_MAX_VERSION;
      |            ^~~~~~~~~~~~~~~
      |            TLS1_2_VERSION
make[2]: *** [Makefile:4470: libnetdata/socket/security.o] Error 1
```

Running `openssl version` returns this:

> OpenSSL 1.1.1g  21 Apr 2020 (Library: OpenSSL 1.1.1d  10 Sep 2019)

### PR DESCRIPTION
**Pull request by thiagoftsm**
##### Summary
Fixes #8796 

This PR fixes the problem to compile Netdata on systems with `libressl`.
##### Component Name
Stream
Web
Exporter
##### Test Plan
I tested this PR installing a VM with Alpine 3.11 and following the steps described on https://github.com/netdata/netdata/issues/8796#issuecomment-617501624

### CODE REVIEW

**Comment by prologic** [comment]
@thiagoftsm Do you mind explaining what's going on here?

Also I don't see any changes to CI that would prevent us from breaking compilation against libressl in future; We should add one. I can do this if you like?

**Comment by thiagoftsm** [comment]
> @thiagoftsm Do you mind explaining what's going on here?
> 
> Also I don't see any changes to CI that would prevent us from breaking compilation against libressl in future; We should add one. I can do this if you like?

It would be great we also have an alpine testing `libressl`.

**Comment by prologic** [comment]
> It would be great we also have an alpine testing `libressl`.

We already do CI against Alpine actually:

```#!sh
$ sift 'alpine' .github/workflows/build-and-install.yml
14:          - 'alpine:edge'
15:          - 'alpine:3.11'
16:          - 'alpine:3.10'
17:          - 'alpine:3.9'
36:          - distro: 'alpine:edge'
38:          - distro: 'alpine:3.11'
40:          - distro: 'alpine:3.10'
42:          - distro: 'alpine:3.9'
```

What we don't do however is test against _any_ Distro __with__ `libressl-dev/devel` installed in place of `openssl-dev/devel`:

```#!sh
$ sift -c libressl packaging/installer/install-required-packages.sh
0
```

The problem I have is that the CI workflow we put in place for testing our ability to build and install the Agent on various supported Distributions of LInux; does not really take into account "customizations". I consider it a user choice to prefer having `libressl` installed in place of `openssl` on a system and I know of no systems that ship by default with `libressl` besides macOS?

**Comment by prologic** [comment]
What's weird is that I'e discovered we "seem" to try to use `libressl-dev` anyway on Debian/Ubuntu (_but I don't believe it_):

```#!sh
$ sift 'ssl-dev' packaging/installer/install-required-packages.sh
978:  ['alpine']="openssl-dev"
979:  ['debian']="libssl-dev"
980:  ['ubuntu']="libssl-dev"
981:  ['suse']="libopenssl-devel"
984:  ['default']="openssl-devel"
```

🤔

**Comment by prologic** [comment]
`libressl-dev` isn't even a package on Debian/Ubuntu.

```#!sh
prologic@Jamess-iMac
Mon Apr 27 12:41:53
~/NetData/netdata
 (master) 0
$ dki -t --rm debian /bin/sh
Unable to find image 'debian:latest' locally
latest: Pulling from library/debian
90fe46dd8199: Pull complete
Digest: sha256:2857989334428416b1ef369d6e029e912a7fe3ee7e57adc20b494cc940198258
Status: Downloaded newer image for debian:latest
# apt-get update
Get:1 http://deb.debian.org/debian buster InRelease [122 kB]
Get:2 http://deb.debian.org/debian buster-updates InRelease [49.3 kB]
Get:3 http://security.debian.org/debian-security buster/updates InRelease [65.4 kB]
Get:4 http://deb.debian.org/debian buster/main amd64 Packages [7907 kB]
Get:5 http://security.debian.org/debian-security buster/updates/main amd64 Packages [189 kB]
Get:6 http://deb.debian.org/debian buster-updates/main amd64 Packages [7380 B]
Fetched 8339 kB in 2s (3854 kB/s)
Reading package lists... Done
# apt-get install libressl-dev
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package libressl-dev
# apt-get install libressl-devel
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package libressl-devel
#

prologic@Jamess-iMac
Mon Apr 27 12:44:37
~/NetData/netdata
 (master) 100
$ dki -t --rm ubuntu /bin/sh
Unable to find image 'ubuntu:latest' locally
latest: Pulling from library/ubuntu
d51af753c3d3: Pull complete
fc878cd0a91c: Pull complete
6154df8ff988: Pull complete
fee5db0ff82f: Pull complete
Digest: sha256:747d2dbbaaee995098c9792d99bd333c6783ce56150d1b11e333bbceed5c54d7
Status: Downloaded newer image for ubuntu:latest
# apt-get update
Get:1 http://security.ubuntu.com/ubuntu focal-security InRelease [97.9 kB]
Get:2 http://archive.ubuntu.com/ubuntu focal InRelease [265 kB]
Get:3 http://archive.ubuntu.com/ubuntu focal-updates InRelease [89.1 kB]
Get:4 http://archive.ubuntu.com/ubuntu focal-backports InRelease [89.2 kB]
Get:5 http://archive.ubuntu.com/ubuntu focal/restricted amd64 Packages [33.4 kB]
Get:6 http://archive.ubuntu.com/ubuntu focal/main amd64 Packages [1275 kB]
Get:7 http://archive.ubuntu.com/ubuntu focal/universe amd64 Packages [11.3 MB]
Get:8 http://archive.ubuntu.com/ubuntu focal/multiverse amd64 Packages [177 kB]
Get:9 http://archive.ubuntu.com/ubuntu focal-updates/main amd64 Packages [1958 B]
Fetched 13.4 MB in 8s (1617 kB/s)
Reading package lists... Done
# apt-get install libressl-dev
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package libressl-dev
# apt-get install libressl-devel
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package libressl-devel
#
```

Confirmed with the folks in #debian on FreeNode. Somehow this is a mistake in our tooling :/

**Comment by prologic** [comment]
I think I need to stop for the day! Apparently my eyes are deceiving me and I mistook `libssl-dev` for `libressl-dev` :/ Ignore my previous two comments (_deleted_).

**Comment by prologic** [comment]
Sorry @thiagoftsm I have no good ideas for how we can introduce CI for libressl right now. I'll have to think some more about it.

**Comment by thiagoftsm** [comment]
> Sorry @thiagoftsm I have no good ideas for how we can introduce CI for libressl right now. I'll have to think some more about it.

No problem @prologic.:) I only used `libressl` on Alpine, but the first time that installed it, I also used `openssl`. Probably to test only with Alpine can be enough.

**Comment by prologic** [comment]
@thiagoftsm #8842 adds a CI check for this; if you could verify that it passes on your PR? :)

**Comment by thiagoftsm** [comment]
@prologic I got the file from #8842 and I notice that there are errors related to the file format https://github.com/netdata/netdata/pull/8807/checks?check_run_id=627498978

**Comment by mandree** [comment]
Is it really worthwhile spending time on a TLS implementation that doesn't support TLS v1.3? I think this is going in the wrong direction. 

And what's with this `LIBRESSL_INTERNAL` in an application that just uses the lib? Looks strange to me.

**Comment by thiagoftsm** [comment]
Hello @mandree ,

> Is it really worthwhile spending time on a TLS implementation that doesn't support TLS v1.3? I think this is going in the wrong direction.

Netdata always keep backward compatibility with old systems that use libraries that did not have  support for the newest TLS versions,  this is the motive we are keeping the code.

> And what's with this `LIBRESSL_INTERNAL` in an application that just uses the lib? Looks strange to me.

Yes, you are right. We are bringing here a possibility for our users to compile using OpenSSL or LibreSSL. For you see the motive the code was changed this way, please, do download of the latest LibreSSL and see the lines 164-167 inside the file `include/openssl/tls1.h.

Best regards!

**Comment by prologic** [comment]
If you fix the failing unit tests and yamllint errors I can approve this :)

**Comment by mandree** [comment]
> > And what's with this `LIBRESSL_INTERNAL` in an application that just uses the lib? Looks strange to me.
> 
> Yes, you are right. We are bringing here a possibility for our users to compile using OpenSSL or LibreSSL. For you see the motive the code was changed this way, please, do download of the latest LibreSSL and see the lines 164-167 inside the file `include/openssl/tls1.h.

Sorry, you are missing my point. `LIBRESSL_INTERNAL` is an internal libressl implementation detail and **does not belong into an application.** 

The attached reduced patch is sufficient to fix compilation with FreeBSD's libressl 3.0.2 port, but still isn't right because it should really test features, not second-guess - for instance, `LIBRESSL_VERSION_NUMBER` or `LIBRESSL_HAS_TLS1_3` are tests that should not be made, and the code should rather check `TLS1_3_VERSION` or `TLS_MAX_VERSION` directly instead. 

I'll not do that because I don't support LibreSSL. If _you_ feel you must, at least don't burden yourself with technical debts.

[libnetdata_socket_security.c.diff.txt](https://github.com/netdata/netdata/files/4555327/libnetdata_socket_security.c.diff.txt)

**Comment by thiagoftsm** [comment]
> If you fix the failing unit tests and yamllint errors I can approve this :)

I think the yaml error came after I bring a file from your PR. I will return to the original.

**Comment by thiagoftsm** [comment]
> [libnetdata_socket_security.c.diff.txt](https://github.com/netdata/netdata/files/4555327/libnetdata_socket_security.c.diff.txt)

I understood better now, thank you for your contribution. I am testing your suggestion and I will bring it case it works.

File: libnetdata/socket/security.c [hunk 1]
```diff
@@ -66,7 +66,7 @@ int tls_select_version(const char *lversion) {
         return TLS1_1_VERSION;
     else if (!strcmp(lversion, "1.2"))
         return TLS1_2_VERSION;
-#if (OPENSSL_VERSION_NUMBER >= OPENSSL_VERSION_111 && !defined(LIBRESSL_VERSION_NUMBER)) || ((defined(LIBRESSL_HAS_TLS1_3) || defined(LIBRESSL_INTERNAL)))
+#if (OPENSSL_VERSION_NUMBER >= OPENSSL_VERSION_111 && !defined(LIBRESSL_VERSION_NUMBER)) || (defined(LIBRESSL_HAS_TLS1_3))
```
Comment (mandree) [inline libnetdata/socket/security.c:69]: Note that also the OPENSSL_* check should check **features** (TLS1_3_VERSION and related) not second-guess from versions. I don't have a vote on the upstream code, but if I did, I'd veto this and request revision.

**Comment by thiagoftsm** [comment]
#9068 brought a solution for this, so I am closing the PR.

**Pull request closed without merging**